tests/test_gemm.py

```diff
@@ -0,0 +1,160 @@
+
+        # Test matrix multiplication
+        result = a.matmul(b)
+
+        self.assertEqual(list(result.shape), [2, 2])
+        np.testing.assert_array_almost_equal(result.ndarray, expected)
+
+    def test_rectangular_matrix_multiplication(self):
+        """Test rectangular matrix multiplication"""
+        # Create 2x3 and 3x2 matrices
+        a_data = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]],
+                          dtype=self.dtype)
+        b_data = np.array([[7.0, 8.0], [9.0, 10.0], [11.0, 12.0]],
+                          dtype=self.dtype)
+
+        a = self.SimpleArray(array=a_data)
+        b = self.SimpleArray(array=b_data)
+
+        # Expected result: [[58, 64], [139, 154]]
+        expected = np.array([[58.0, 64.0], [139.0, 154.0]],
+                            dtype=self.dtype)
+
+        result = a.matmul(b)
+
+        self.assertEqual(list(result.shape), [2, 2])
+        np.testing.assert_array_almost_equal(result.ndarray, expected)
+
+    def test_identity_matrix(self):
+        """Test multiplication with identity matrix"""
+        # 3x3 matrix
+        a_data = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0],
+                           [7.0, 8.0, 9.0]], dtype=self.dtype)
+        identity_data = np.eye(3, dtype=self.dtype)
+
+        a = self.SimpleArray(array=a_data)
+        identity = self.SimpleArray(array=identity_data)
+
+        result = a.matmul(identity)
+
+        self.assertEqual(list(result.shape), [3, 3])
+        np.testing.assert_array_almost_equal(result.ndarray, a_data)
+
+    def test_zero_matrix(self):
+        """Test multiplication with zero matrix"""
+        a_data = np.array([[1.0, 2.0], [3.0, 4.0]], dtype=self.dtype)
+        zero_data = np.zeros((2, 2), dtype=self.dtype)
+
+        a = self.SimpleArray(array=a_data)
+        zero = self.SimpleArray(array=zero_data)
+
+        result = a.matmul(zero)
+
+        self.assertEqual(list(result.shape), [2, 2])
+        np.testing.assert_array_almost_equal(result.ndarray, zero_data)
+
+    def test_dimension_mismatch_error(self):
+        """Test error handling for incompatible dimensions"""
+        a_data = np.array([[1.0, 2.0], [3.0, 4.0]],
+                          dtype=self.dtype)  # 2x2
+        b_data = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0],
+                           [7.0, 8.0, 9.0]], dtype=self.dtype)  # 3x3
+
+        a = self.SimpleArray(array=a_data)
+        b = self.SimpleArray(array=b_data)
+
+        # Should raise error: 2x2 cannot multiply with 3x3
+        # (incompatible inner dimensions)
+        with self.assertRaises(RuntimeError):
+            a.matmul(b)
+
+    def test_compare_with_numpy(self):
+        """Compare results with NumPy for various matrix sizes"""
+        sizes = [(2, 3, 4), (5, 5, 5), (4, 6, 3)]
+
+        for m, k, n in sizes:
+            with self.subTest(m=m, k=k, n=n):
+                # Generate random matrices
+                np.random.seed(42)  # For reproducible results
```

Comment:
Do not generate random numbers when testing because it removes the fixture (numbers) from the testing code.  The explicit numbers serve as document and examples.

Generate the numbers separately and hard-code them in the testing code as fixture instead.